internal/reminder/reminder.go

```diff
@@ -149,70 +139,174 @@ func (r *reminder) Stop() {
+
+	intersectionPoint := -1
+	var additionalRepos []db.Repository
+
+	// Only fetch additional repositories if we are under the limit
+	if len(repos) < r.cfg.RecurrenceConfig.BatchSize {
+		additionalRepos, err = r.getAdditionalRepos(ctx, repos)
+		if err != nil {
+			return nil, err
+		}
+
+		repos, intersectionPoint = r.mergeRepoBatch(repos, additionalRepos)
 	}
 
-	return os.WriteFile(r.cfg.CursorFile, buf.Bytes(), 0600)
+	r.updateRepositoryCursor(ctx, repos, additionalRepos, intersectionPoint)
+
+	return repos, nil
 }
 
-// restoreCursorState restores the cursor state from a file
-// Not thread-safe, should be called from a single goroutine
-func (r *reminder) restoreCursorState(ctx context.Context) error {
+func (r *reminder) getAdditionalRepos(ctx context.Context, alreadyFetchedRepos []db.Repository) ([]db.Repository, error) {
 	logger := zerolog.Ctx(ctx)
-	logger.Debug().Msg("restoring cursor state")
 
-	if _, err := os.Stat(r.cfg.CursorFile); os.IsNotExist(err) {
-		return nil
+	if len(alreadyFetchedRepos) == 0 {
+		logger.Debug().Msgf("no repositories found after cursor %s, resetting cursor to zero uuid", r.repositoryCursor)
+		r.repositoryCursor = uuid.Nil
+	} else {
+		r.repositoryCursor = alreadyFetchedRepos[len(alreadyFetchedRepos)-1].ID
+		r.adjustCursorForEndOfList(ctx)
 	}
 
-	data, err := os.ReadFile(r.cfg.CursorFile)
+	remaining := r.cfg.RecurrenceConfig.BatchSize - len(alreadyFetchedRepos)
+	duration := pgtype.Interval{}
+	err := duration.Set(r.cfg.RecurrenceConfig.MinElapsed)
 	if err != nil {
-		return err
+		return nil, err
 	}
 
-	buf := bytes.NewBuffer(data)
-	dec := gob.NewDecoder(buf)
-
-	cursorData := make(map[string]interface{})
+	logger.Debug().Msgf("fetching additional repositories after cursor: %s", r.repositoryCursor)
+	repos, err := r.store.ListEligibleRepositoriesAfterID(ctx, db.ListEligibleRepositoriesAfterIDParams{
+		ID:         r.repositoryCursor,
+		MinElapsed: duration,
+		Limit: sql.NullInt64{
+			Int64: int64(remaining),
+			Valid: true,
+		},
+	})
+	if err != nil {
+		return nil, err
+	}
 
-	if err := dec.Decode(&cursorData); err != nil {
-		return err
+	if len(repos) == 0 {
+		logger.Debug().Msg("no additional repositories fetched")
+	} else {
+		logger.Debug().Msgf("fetched %d additional repositories", len(repos))
 	}
 
-	if val, ok := cursorData["projectListCursor"]; ok {
-		if v, ok := val.(string); ok {
-			r.projectListCursor = v
-		} else {
-			return fmt.Errorf("projectListCursor is not a string")
+	return repos, nil
+}
+
+func (_ *reminder) mergeRepoBatch(repos []db.Repository, additionalRepos []db.Repository) ([]db.Repository, int) {
+	var mergedBatch []db.Repository
+	mergedBatch = append(mergedBatch, repos...)
+	reposSet := sets.New[db.Repository](repos...)
+
+	// There may be an intersection between the two sets
+	// If there is an intersection, then we need to update the cursor to the last fetched
+	// non-common repository
```

Comment:
I'm not sure I understand how this happens, unless there are only e.g. 4 repos eligible in the whole database.